blog/content/post/2025-04-10-CROSSCON-its-Hypervisor-and-Zarhus.md

```diff
@@ -0,0 +1,481 @@
+[    0.000000] arch_timer: cp15 timer(s) running at 54.00MHz (virt).
+[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0xc743ce346, max_idle_ns: 440795203123 ns
+[    0.000000] sched_clock: 56 bits at 54MHz, resolution 18ns, wraps every 4398046511102ns
+[    0.008334] Console: colour dummy device 80x25
+[    0.012844] printk: console [tty0] enabled
+[    0.017000] printk: bootconsole [bcm2835aux0] disabled
+```
+
+but that was the end of the output - it seemed to freeze. Once I noticed where
+it was freezing:
+
+```logs
+[    0.017000] printk: bootconsole [bcm2835aux0] disabled
+```
+
+I knew there was some sort of serial console issue. I suspected that the system
+was booting normally and without errors, and just not printing the output
+because of an unconfigured console.
+
+Adding this `console=ttyS1,115200` to `bootargs` in the device tree file used in
+[step 9](https://github.com/crosscon/CROSSCON-Hypervisor-and-TEE-Isolation-Demos/tree/master/rpi4-ws#step-9-bind-linux-image-and-device-tree)
+fixed the issue, but another one arose:
+
+```logs
+[    1.953218] bcm2835-aux-uart fe215040.serial: error -EINVAL: unable to register 8250 port
+```
+
+This is was an easy fix, again adding to `bootargs`, but this time
+`8250.nr_uarts=8`. This line tells the `8250` serial driver to allocate up to 8
+ports - the number doesn't really matter that much here, but by default it is
+one, and that's not enough for the serial setup that we have.
+
+### Mounting the rootfs
+
+It was only after fixing the serial console issues, that I could uncover the
+real issues - the kernel was panicking after all, I just couldn't see it
+because of the lack of serial output:
+
+```logs
+[    2.898654] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
+[    2.907033] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 6.6.22-v8 #1
+[    2.913298] Hardware name: Raspberry Pi 4 Model B (DT)
+[    2.918502] Call trace:
+[    2.920974]  dump_backtrace+0x9c/0x100
+[    2.924776]  show_stack+0x20/0x38
+[    2.928133]  dump_stack_lvl+0x48/0x60
+[    2.931842]  dump_stack+0x18/0x28
+[    2.935197]  panic+0x328/0x390
+[    2.938291]  mount_root_generic+0x26c/0x348
+[    2.942530]  mount_root+0x17c/0x348
+[    2.946062]  prepare_namespace+0x74/0x2b8
+[    2.950123]  kernel_init_freeable+0x374/0x3d8
+[    2.954537]  kernel_init+0x2c/0x1f8
+[    2.958070]  ret_from_fork+0x10/0x20
+[    2.961693] Kernel Offset: 0x80000 from 0xffffffc080000000
+[    2.967250] PHYS_OFFSET: 0x0
+[    2.970162] CPU features: 0x0,80000200,3c020000,0000421b
+[    2.975543] Memory Limit: none
+[    2.978634] ---[ end Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0) ]---
+```
+
+This was an oversight on my part - the rootfs was definitely there on the second
+partition, but I wasn't telling the kernel properly on where it was. Because of
+this, the kernel couldn't find it, and it panicked.
```

Comment:
This paragraph could be more to the point. Also, reading too much "this" and "that" is hard. I don't want to destroy your style, but it also has to be consumable content, so I would appreciate some work, at least a little bit, to eliminate such pointers.